src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs

```diff
@@ -621,6 +622,44 @@ public ValueTask HandleAsync(ProcessAuthenticationContext context)
         }
     }
 
+    /// <summary>
+    /// Contains the logic responsible for overriding the scope to be attached
+    /// to the token request for the providers that require it.
+    /// </summary>
+    public sealed class OverrideScope : IOpenIddictClientHandler<ProcessAuthenticationContext>
+    {
+        /// <summary>
+        /// Gets the default descriptor definition assigned to this handler.
+        /// </summary>
+        public static OpenIddictClientHandlerDescriptor Descriptor { get; }
+            = OpenIddictClientHandlerDescriptor.CreateBuilder<ProcessAuthenticationContext>()
+                .AddFilter<RequireTokenRequest>()
+                .UseSingletonHandler<OverrideScope>()
+                .SetOrder(AttachTokenRequestParameters.Descriptor.Order - 500)
+                .SetType(OpenIddictClientHandlerType.BuiltIn)
+                .Build();
+
+        /// <inheritdoc />
+        public ValueTask HandleAsync(ProcessAuthenticationContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // osu! requires that the scope must be "public" for client credentials
+            // grant, as other scopes have no meaningful effect.
+            if (context.GrantType is GrantTypes.ClientCredentials &&
+                context.Registration.ProviderType is ProviderTypes.Osu)
+            {
+                context.Scopes.Add("public");
+                // TODO: how does user code add custom scopes here?
```

Comment:
As mentioned in my other reply, scopes are added by the developers via `ClientCredentialsAuthenticationRequest.Scopes`. When doing so, these scopes are automatically attached to `ProcessAuthenticationContext.Scopes` by OpenIddict.